src/openslide-vendor-dicom.c

```diff
@@ -0,0 +1,943 @@
+                      _openslide_format_double(mmpp * 1000));
+}
+
+static void file_hash(const struct dicom_file *f,
+                      struct _openslide_hash *quickhash1) {
+  uint32_t n = dcm_dataset_count(f->metadata);
+  for(uint32_t i = 0; i < n; i++) {
+      DcmElement *element = dcm_dataset_get(NULL, f->metadata, i);
+      if (element) {
+        uint32_t tag = dcm_element_get_tag(element);
+        _openslide_hash_data(quickhash1, &tag, sizeof(tag));
+        // need to get the vr class
+        //_openslide_hash_data(quickhash1, data, datalen);
+      }
+  }
+}
+
+static bool dicom_open(openslide_t *osr, 
+                       const char *filename,
+                       struct _openslide_tifflike *tl G_GNUC_UNUSED,
+                       struct _openslide_hash *quickhash1, 
+                       GError **err) {
+  g_autofree char *dirname = g_path_get_dirname(filename);
+
+  g_autoptr(GDir) dir = g_dir_open(dirname, 0, err);
+  if (!dir) {
+    return false;
+  }
+
+  g_autoptr(GHashTable) dicom_file_hash =
+    g_hash_table_new_full(g_str_hash, 
+                          g_str_equal, 
+                          g_free,
+                          (GDestroyNotify) dicom_file_destroy);
+
+  // open all DICOM files that look like parts of a slide image and 
+  // get the file metadata
+  const char *name;
+  while ((name = g_dir_read_name(dir))) {
+    char *filename = g_build_path("/", dirname, name, NULL);
+
+#ifdef DEBUG
+    printf("trying to open: %s ...\n", filename);
+#endif /*DEBUG*/
+    GError *local_error = NULL;
+    struct dicom_file *f = dicom_file_new(filename, &local_error);
+    if (!f) {
+      g_free(filename);
+#ifdef DEBUG
+      printf( "open failed: %s\n", local_error->message);
+      g_error_free(local_error);
+#endif /*DEBUG*/
+      continue;
+    }
+
+    g_hash_table_insert(dicom_file_hash, filename, f);
+  }
+
+#ifdef DEBUG
+  printf("found WSI DICOM files:\n");
+  g_hash_table_foreach(dicom_file_hash, print_file_from_hash, NULL);
+#endif /*DEBUG*/
+
+  // pull out the subset of DICOM files that look like pyramid levels
+  g_autoptr(GHashTable) level_hash =
+    g_hash_table_new_full(g_str_hash,
+                          g_str_equal, 
+                          g_free,
+                          (GDestroyNotify) level_destroy);
+  g_hash_table_foreach_steal(dicom_file_hash, find_levels, level_hash);
+
+#ifdef DEBUG
+  printf("found pyr levels DICOM files:\n");
+  g_hash_table_foreach(level_hash, print_level_from_hash, NULL);
+#endif /*DEBUG*/
+
+  // we can have several slides in one directory -- find the largest pyramid
+  // layer and pick that as the slide image we are opening
+  const struct level *largest = NULL;
+  g_hash_table_foreach(level_hash, find_largest, &largest);
+  if (!largest) {
+    return false;
+  }
+
+  const char *slide_id;
+  if (!get_tag_str(largest->file->metadata, 
+                   "SeriesInstanceUID", 
```

Comment:
Let's define all of these tags as `static const char[]` at the top of the file.

Reply:
OK, done! I didn't use a prefix on the var names, hopefully that's OK.